packages/react/src/components/TextArea/TextArea.tsx

```diff
@@ -155,31 +161,103 @@ const TextArea = React.forwardRef((props: TextAreaProps, forwardRef) => {
   const prefix = usePrefix();
   const { isFluid } = useContext(FormContext);
   const { defaultValue, value, disabled } = other;
+
+  function getInitialTextCount(
+    value: string | number | undefined,
+    defaultValue: string | number | undefined,
+    counterMode: 'character' | 'word' | undefined
+  ): number {
+    if (defaultValue) {
+      if (counterMode === 'character') {
+        return defaultValue.toString().length;
+      } else {
+        return defaultValue.toString().match(/\w+/g)?.length || 0;
+      }
+    } else if (value) {
+      if (counterMode === 'character') {
+        return value.toString().length;
+      } else {
+        return value.toString().match(/\w+/g)?.length || 0;
+      }
+    }
+    return 0;
+  }
+
   const [textCount, setTextCount] = useState(
-    defaultValue?.toString()?.length || value?.toString()?.length || 0
+    getInitialTextCount(defaultValue, value, counterMode)
   );
   const { current: textAreaInstanceId } = useRef(getInstanceId());
 
   useEffect(() => {
     setTextCount(
-      defaultValue?.toString()?.length || value?.toString()?.length || 0
+      defaultValue
+        ? counterMode === 'character'
+          ? defaultValue.toString().length
+          : defaultValue.toString().match(/\w+/g)?.length || 0
+        : value
+        ? counterMode === 'character'
+          ? value.toString().length
+          : value.toString().match(/\w+/g)?.length || 0
+        : 0
```

Comment:
can we call the getinitialtextcount function here as well?

Reply:
My bad must have missed it